### PR TITLE
feat(core): make the `upgrade` module an implementation detail

### LINKED ISSUE #3748: core: make `upgrade::apply` an implementation detail of transport upgrades
## Description

Currently, we expose the generic `upgrade::apply` function. I'd suggest we make it an implementation detail and transition all usages to something else. The usages are only in tests and in `from_fn` (which I am suggesting to remove: https://github.com/libp2p/rust-libp2p/pull/3747).

## Motivation

Smaller, more focused API surface.

## Are you planning to do it yourself in a pull request?

Yes.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Resolves: #3748.

## Notes & open questions

~Draft because stacked.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/CHANGELOG.md [hunk 1]
```diff
@@ -19,10 +19,15 @@
...
 
+- Remove `upgrade::{apply, apply_inbound, apply_outbound, InboundUpgradeApply, OutboundUpgradeApply}` from public API.
+  These are implementation details that should not be depended on.
+  See [PR XXXX].
+
 [spec]: https://github.com/libp2p/specs/blob/master/connections/README.md#multistream-select
 [PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
 [PR 3883]: https://github.com/libp2p/rust-libp2p/pull/3883
 [PR 3814]: https://github.com/libp2p/rust-libp2p/pull/3814
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline core/CHANGELOG.md]:
```suggestion
  See [PR 3915].

[spec]: https://github.com/libp2p/specs/blob/master/connections/README.md#multistream-select
[PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
[PR 3883]: https://github.com/libp2p/rust-libp2p/pull/3883
[PR 3814]: https://github.com/libp2p/rust-libp2p/pull/3814
[PR 3915]: https://github.com/libp2p/rust-libp2p/pull/3915
```

**Comment by thomaseizinger** [comment]
In a later step, we can simplify some of this code down to use boxed futures which should save us quite a bit of code!

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
Can you remove the usage in `libp2p-plaintext` @thomaseizinger?

See CI failure:

```
---- src/lib.rs - PlainText1Config (line 58) stdout ----
error[E0603]: function `apply` is private
error: doctest failed, to rerun pass `-p libp2p-plaintext --doc`
Error:   --> src/lib.rs:64:27
   |
9  |     libp2p_core::upgrade::apply(
   |                           ^^^^^ private function
   |
note: the function `apply` is defined here
  --> /home/runner/work/rust-libp2p/rust-libp2p/core/src/upgrade.rs:70:5
   |
70 |     apply, apply_inbound, apply_outbound, InboundUpgradeApply, OutboundUpgradeApply,
   |     ^^^^^

error: aborting due to previous error
```

**Comment by thomaseizinger** [comment]
> Can you remove the usage in `libp2p-plaintext` @thomaseizinger?

I thought I had removed all usages before! Fixed now :)

**Comment by mxinden** [comment]
@thomaseizinger `libp2p-plaintext` tests still failing:

```
---- src/lib.rs - PlainText1Config (line 58) stdout ----
error[E0271]: type mismatch resolving `<PlainText1Config as InboundUpgrade<Negotiated<RwStreamSink<Chan>>>>::Output == (PeerId, _)`
Error:    --> src/lib.rs:64:17
    |
9   |   .authenticate(PlainText1Config {})
    |    ------------ ^^^^^^^^^^^^^^^^^^^ expected `(PeerId, _)`, found `Negotiated<RwStreamSink<Chan>>`
    |    |
    |    required by a bound introduced by this call
    |
    = note: expected tuple `(libp2p_identity::peer_id::PeerId, _)`
              found struct `multistream_select::negotiated::Negotiated<rw_stream_sink::RwStreamSink<Chan>>`
note: required by a bound in `libp2p_core::transport::upgrade::Builder::<T>::authenticate`
   --> /home/runner/work/rust-libp2p/rust-libp2p/core/src/transport/upgrade.rs:104:42
    |
104 |         U: InboundUpgrade<Negotiated<C>, Output = (PeerId, D), Error = E>,
    |                                          ^^^^^^^^^^^^^^^^^^^^ required by this bound in `Builder::<T>::authenticate`

error[E0271]: type mismatch resolving `<PlainText1Config as OutboundUpgrade<Negotiated<RwStreamSink<Chan>>>>::Output == (PeerId, _)`
```

**Comment by thomaseizinger** [comment]
Yeah I noticed. Should we just bin `Plaintext1Config`? It is not compatible with the upgrade API so why would anyone use that at all?

**Comment by mxinden** [comment]
Removing `Plaintext1Config` sounds good to me. I don't see a use-case where `Plaintext2Config` is not enough.

**Comment by thomaseizinger** [comment]
> Removing `Plaintext1Config` sounds good to me. I don't see a use-case where `Plaintext2Config` is not enough.

Opened a PR here: https://github.com/libp2p/rust-libp2p/pull/3940

**Pull request merged**